core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -298,11 +294,9 @@ final class ShuffleBlockFetcherIterator(
         // not exist, SPARK-4085). In that case, we should propagate the right exception so
         // the scheduler gets a FetchFailedException.
         Try(buf.createInputStream()).map { is0 =>
-          val is = blockManager.wrapForCompression(blockId, is0)
-          val iter = serializerInstance.deserializeStream(is).asKeyValueIterator
-          CompletionIterator[Any, Iterator[Any]](iter, {
-            // Once the iterator is exhausted, release the buffer and set currentResult to null
-            // so we don't release it again in cleanup.
+          // Once the single-element (is0) iterator is exhausted, release the buffer so that we
+          // don't release it again in cleanup.
+          CompletionIterator[InputStream, Iterator[InputStream]](Iterator(is0), {
```

Comment:
I like the idea of returning an `InputStream` that requires callers to call `close()` and drop the use of the single-element CompletionIterator. I'll take make that change now.